torch/fx/graph.py

```diff
@@ -952,6 +952,20 @@ def check_arg(arg : Node, n : Optional[Node] = None) -> None:
                             raise RuntimeError(f'Node {node} target {node.target} references nonexistent attribute '
                                                f'{atom} of {seen_qualname}')
 
+    def run_dce(self):
+        """
+        Remove all dead code from the graph, based on each node's number of users.
+        """
+        # Lint the graph first to make sure its topologically sorted, otherwise
+        # DCE below will not behave as expected.
+        self.lint()
+        # Reverse iterate so that when we remove a node, any nodes used as an
+        # input to that node have an updated user count that no longer reflects
+        # the removed node.
+        for node in reversed(self.nodes):
+            if node.op not in {"placeholder", "output"} and len(node.users) == 0:
```

Comment:
Let's move this condition out into a method `Node.is_impure`. That way we can have a single, overridable definition for impurity of nodes that other passes can utilize as well

Reply:
Cool, so I've added it -- but I also added another quirk here in the latest update. Specifically, some other nodes may have side effects which means we should not remove them even if they have no users.

This may be universal (i.e. regardless of use case), for example for `call_function` to `torch._assert`, and so could be added to `is_impure()` in a straightforward way.

But it may also be use case dependent -- for example if we have some special Module that we don't trace into but we know it has its own asserts or raises exceptions, and so should not be eliminated.

I added more special handling for these cases for now inside `eliminate_dead_code()` directly, with extra args passed in to handle them, but would be great to get your feedback on if it's handled in the best way -- e.g. perhaps we should move those to `Node.is_impure()`? Or perhaps something else.